src/language-services/documentSymbols.ts

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as PQP from "@microsoft/powerquery-parser";
+
+import { AnalysisOptions } from "./analysisOptions";
+import { DocumentSymbol, Range, SymbolKind, TextDocument } from "./commonTypes";
+import * as InspectionUtils from "./inspectionUtils";
+import * as LanguageServiceUtils from "./languageServiceUtils";
+import * as WorkspaceCache from "./workspaceCache";
+
+const DefaultLocale: string = PQP.Locale.en_US;
+
+export function getDocumentSymbols(document: TextDocument, options?: AnalysisOptions): DocumentSymbol[] {
+    const triedLexParse: PQP.Task.TriedLexParse = WorkspaceCache.getTriedLexParse(document);
+    let result: DocumentSymbol[] = [];
+
+    // TODO: Can we get symbols even when there is a syntax error?
+    if (PQP.ResultUtils.isOk(triedLexParse)) {
+        const lexParseOk: PQP.Task.LexParseOk = triedLexParse.value;
+        const ast: PQP.Language.Ast.TNode = lexParseOk.ast;
+        const nodeIdMapCollection: PQP.NodeIdMap.Collection = lexParseOk.state.contextState.nodeIdMapCollection;
+
+        const documentOutlineResult: PQP.Traverse.TriedTraverse<DocumentOutline> = tryTraverse(
+            ast,
+            nodeIdMapCollection,
+            options,
+        );
+
+        // TODO: Trace error case
+        if (PQP.ResultUtils.isOk(documentOutlineResult)) {
+            result = documentOutlineResult.value.symbols;
+        }
+    }
+
+    if (!options?.maintainWorkspaceCache) {
+        WorkspaceCache.close(document);
+    }
+
+    return result;
+}
+
+interface DocumentOutline {
+    symbols: DocumentSymbol[];
+}
+
+interface TraversalState extends PQP.Traverse.IState<DocumentOutline> {
+    readonly nodeIdMapCollection: PQP.NodeIdMap.Collection;
+    parentSymbolMap: Map<number, DocumentSymbol>;
+}
+
+function tryTraverse(
+    ast: PQP.Language.Ast.TNode,
+    nodeIdMapCollection: PQP.NodeIdMap.Collection,
+    options?: AnalysisOptions,
+): PQP.Traverse.TriedTraverse<DocumentOutline> {
+    const locale: string = options?.locale ?? DefaultLocale;
+    const localizationTemplates: PQP.ILocalizationTemplates = PQP.getLocalizationTemplates(locale);
+
+    const traversalState: TraversalState = {
+        localizationTemplates,
+        nodeIdMapCollection,
+        parentSymbolMap: new Map<number, DocumentSymbol>(),
+        result: {
+            symbols: [],
+        },
+    };
+
+    return PQP.Traverse.tryTraverseAst(
+        traversalState,
+        nodeIdMapCollection,
+        ast,
+        PQP.Traverse.VisitNodeStrategy.BreadthFirst,
+        visitNode,
+        PQP.Traverse.expectExpandAllAstChildren,
+        undefined,
```

Comment:
Can you add a function here? It can hugely speed up a traversal. Looks like anything that can't be/contain either a Section/Record/IdentifierPairedExpression can be ignored.

Reply:
I added the function, but had a hard time figuring out where to stop the traversal.